### PR TITLE
[JENKINS-59959] Fix concurrency limits

### PR DESCRIPTION
**Pull request by schottsfired**
- Modified provisioning logic so that concurrency limits are always respected
- General clean-up and updated names
- Make `test-in-k8s.sh` executable without `sudo`

### CODE REVIEW

File: src/main/java/org/csanchez/jenkins/plugins/kubernetes/KubernetesCloud.java [hunk 1]
```diff
@@ -580,9 +579,10 @@ public KubernetesClient connect() throws KubernetesAuthException, IOException {
      * Check not too many already running.
      *
      */
-    private boolean addProvisionedSlave(@Nonnull PodTemplate template, @CheckForNull Label label, int scheduledCount) throws Exception {
+    private boolean addProvisionedSlave(@Nonnull PodTemplate template, @CheckForNull Label label, int numProvisioned,
+                                        int numInProvisioning) throws Exception {
         if (containerCap == 0) {
-            return true;
+            return false;
```
Comment (Vlatombe) [inline src/main/java/org/csanchez/jenkins/plugins/kubernetes/KubernetesCloud.java:585]: Should be in another PR

File: src/main/java/org/csanchez/jenkins/plugins/kubernetes/KubernetesCloud.java [hunk 2]
```diff
@@ -593,23 +593,27 @@ private boolean addProvisionedSlave(@Nonnull PodTemplate template, @CheckForNull
...
+        // check template-level concurrency limit using template-level labels
+        Map<String, String> templateLabels = new HashMap<>(podLabels);
+        templateLabels.putAll(template.getLabelsMap());
+        numRunningOrPending = getNumActiveSlavePods(client, templateNamespace, podLabels);
+        if (numRunningOrPending + numProvisioned + numInProvisioning >= template.getInstanceCap()) {
             LOGGER.log(Level.INFO,
-                    "Maximum number of concurrently running agent pods ({0}) reached for template {1} in Kubernetes Cloud {6}, not provisioning: {2} running or pending in namespace {3} with label \"{4}\" and Kubernetes labels {5}",
-                    new Object[] { template.getInstanceCap(), template.getName(), activeTemplateSlavePods.size() + scheduledCount,
-                            templateNamespace, label == null ? "" : label.toString(), labelsMap, name });
+                    "Maximum number of concurrently running agent pods ({0}) reached for template {1} in Kubernetes Cloud {6}, " +
```
Comment (Vlatombe) [inline src/main/java/org/csanchez/jenkins/plugins/kubernetes/KubernetesCloud.java:613]: Avoid reformatting if you can to make the review easier.

**Review by Vlatombe — CHANGES_REQUESTED** [review]
This breaks provisioning from other clouds if the requested label can't be served by any kubernetes cloud.

The key problem here is that in order to apply the correct limit, the `KubernetesCloud` would need access to the planned capacity snapshot which is currently stored in `StrategyState`, and therefore only visible from the `Strategy`.

File: src/main/java/io/jenkins/plugins/kubernetes/NoDelayProvisionerStrategy.java [hunk 3]
```diff
@@ -30,12 +30,12 @@
...
 
     private static final Logger LOGGER = Logger.getLogger(NoDelayProvisionerStrategy.class.getName());
-    private static final boolean DISABLE_NODELAY_PROVISING = Boolean.valueOf(
+    private static final boolean DISABLE_NO_DELAY_PROVISIONING = Boolean.parseBoolean(
             System.getProperty("io.jenkins.plugins.kubernetes.disableNoDelayProvisioning"));
 
     @Override
     public NodeProvisioner.StrategyDecision apply(NodeProvisioner.StrategyState strategyState) {
-        if (DISABLE_NODELAY_PROVISING) {
+        if (DISABLE_NO_DELAY_PROVISIONING) {
```
Comment (jglick) [inline src/main/java/io/jenkins/plugins/kubernetes/NoDelayProvisionerStrategy.java:38]: Keep the diff smaller by avoiding gratuitous changes. Can always file separate stylistic cleanup PRs.

File: src/main/java/org/csanchez/jenkins/plugins/kubernetes/KubernetesCloud.java [hunk 4]
```diff
@@ -541,24 +539,25 @@ public KubernetesClient connect() throws KubernetesAuthException, IOException {
             int toBeProvisioned = Math.max(0, excessWorkload - allInProvisioning.size());
             LOGGER.log(Level.INFO, "Excess workload after pending Kubernetes agents: {0}", toBeProvisioned);
 
-            List<NodeProvisioner.PlannedNode> r = new ArrayList<NodeProvisioner.PlannedNode>();
+            List<NodeProvisioner.PlannedNode> plannedNodes = new ArrayList<>();
 
-            for (PodTemplate t: getTemplatesFor(label)) {
-                LOGGER.log(Level.INFO, "Template for label {0}: {1}", new Object[] { label, t.getName() });
+            for (PodTemplate podTemplate: getTemplatesFor(label)) {
+                LOGGER.log(Level.INFO, "Template for label {0}: {1}", new Object[] { label, podTemplate.getName() });
```
Comment (jglick) [inline src/main/java/org/csanchez/jenkins/plugins/kubernetes/KubernetesCloud.java:545]: Also could be reverted to minimize review surface.

**Comment by schottsfired** [comment]
Thanks for your reviews @Vlatombe and @jglick. Apologies for the extra code, the changes helped me understand what was what while getting closer to a fix.

For context, the key issues that caused overprovisioning were:
1. The K8s provisioner strategy consulted other strategies too often, e.g. immediately after handling provisioning but hitting a concurrency limit.
2. The concurrency limit check did not account for `numInProvisioning`.

https://github.com/jenkinsci/kubernetes-plugin/pull/824/commits/dd32d49c7e24be7334f8cc412e70e779b26d8c63 should take care of the breakage mentioned by @Vlatombe in https://github.com/jenkinsci/kubernetes-plugin/pull/824#pullrequestreview-477671568.

**Comment by schottsfired** [comment]
Hmm, no luck kicking CI.

While I'm here I'll share the Pipeline script used for testing:
```
properties([
    parameters([
        string(name: 'numParallel', 
        description: 'number of parallel agents to run',
        defaultValue: '6')
    ])
])
def steps = [failFast: false]
for (int i = 0; i < params.numParallel.toInteger(); i++) {
    def buildNum = i
    steps["build ${buildNum}"] = {
        stage ("build stuff that matters ${buildNum}") {
            node('testlabel') {
                def seconds = 15 * Math.random()
                sh "sleep ${seconds}"
            }
        }
    }
}
parallel steps
```

File: src/main/java/org/csanchez/jenkins/plugins/kubernetes/KubernetesCloud.java (hunk 1 above)
Comment (Vlatombe) [reply src/main/java/org/csanchez/jenkins/plugins/kubernetes/KubernetesCloud.java:585]: Please move this one to another PR for clarity.

File: src/test/java/org/csanchez/jenkins/plugins/kubernetes/KubernetesCloudTest.java [hunk 5]
```diff
@@ -221,6 +221,11 @@ public KubernetesClient connect()  {
         Collection<NodeProvisioner.PlannedNode> plannedNodes = cloud.provision(test, 200);
         assertEquals(200, plannedNodes.size());
 
+        cloud.setContainerCapStr("0");
```
Comment (Vlatombe) [inline src/test/java/org/csanchez/jenkins/plugins/kubernetes/KubernetesCloudTest.java:224]: I guess it goes with https://github.com/jenkinsci/kubernetes-plugin/pull/824/files#r479202733 ?

Comment (schottsfired) [reply src/test/java/org/csanchez/jenkins/plugins/kubernetes/KubernetesCloudTest.java:224]: yes, it does

File: src/main/java/org/csanchez/jenkins/plugins/kubernetes/KubernetesCloud.java (hunk 1 above)
Comment (schottsfired) [reply src/main/java/org/csanchez/jenkins/plugins/kubernetes/KubernetesCloud.java:585]: OK, will do.

Comment (schottsfired) [reply src/main/java/org/csanchez/jenkins/plugins/kubernetes/KubernetesCloud.java:585]: Moved to https://github.com/jenkinsci/kubernetes-plugin/pull/828

**Comment by schottsfired** [comment]
The fix for "concurrency limit: 0" has been moved to https://github.com/jenkinsci/kubernetes-plugin/pull/828.

File: src/main/java/org/csanchez/jenkins/plugins/kubernetes/KubernetesCloud.java [hunk 6]
```diff
@@ -541,24 +539,25 @@ public KubernetesClient connect() throws KubernetesAuthException, IOException {
...
                 for (int i = 0; i < toBeProvisioned; i++) {
-                    if (!addProvisionedSlave(t, label, i)) {
+                    // Check concurrency limits
+                    if (!addProvisionedSlave(podTemplate, label, i, allInProvisioning.size())) {
```
[the remark this answers is not in the document]
Comment (schottsfired) [reply src/main/java/org/csanchez/jenkins/plugins/kubernetes/KubernetesCloud.java:548]: Right, this is a better place to do the addition vs. modifying the method signature.

**Comment by Vlatombe** [comment]
Looks like you changed permissions of `test-in-k8s.sh`. Could you revert it ?

**Comment by schottsfired** [comment]
@Vlatombe I don't think so, unless I'm somehow using `git clone` improperly, which I doubt.

I'm glad you brought it up. I changed permissions to `755` (executable by everyone), however `744` should suffice.

Here it is in a different branch:
```
drwxr-xr-x   4 schottsfired  staff     128 May 30  2019 src
drwxr-xr-x  16 schottsfired  staff     512 Aug 28 09:30 target
-rw-r--r--   1 schottsfired  staff    1259 Aug 31 14:17 test-in-k8s.sh
-rw-r--r--   1 schottsfired  staff    3303 May 15 10:19 test-in-k8s.yaml
drwxr-xr-x  40 schottsfired  staff    1280 Aug 28 09:30 work
➜  kubernetes-plugin git:(dind-example-sans-tls) ./test-in-k8s.sh
zsh: permission denied: ./test-in-k8s.sh
➜  kubernetes-plugin git:(dind-example-sans-tls) sudo chmod 744 test-in-k8s.sh
Password:
➜  kubernetes-plugin git:(dind-example-sans-tls) ✗ ./test-in-k8s.sh
+ kubectl get ns kubernetes-plugin-test
NAME                     STATUS   AGE
kubernetes-plugin-test   Active   334d
+ kubectl get ns kubernetes-plugin-test-overridden-namespace
```

**Comment by Vlatombe** [comment]
Ah, the CI uses `bash test-in-k8s.sh` so it doesn't need it to be executable. Personally, I don't use that script (I test against docker desktop), but I don't mind changing it.

File: src/main/java/io/jenkins/plugins/kubernetes/NoDelayProvisionerStrategy.java [hunk 7]
```diff
@@ -64,26 +65,29 @@
...
-            LOGGER.log(Level.FINE, "Provisioning completed");
-            return NodeProvisioner.StrategyDecision.PROVISIONING_COMPLETED;
+        if (currentDemand - availableCapacity <= 0) {
+            LOGGER.log(Level.FINE, String.format("Provisioning completed for label: [%s]", label));
         } else {
-            LOGGER.log(Level.FINE, "Provisioning not complete, consulting remaining strategies");
-            return NodeProvisioner.StrategyDecision.CONSULT_REMAINING_STRATEGIES;
+            if (!canProvision) {
+                return NodeProvisioner.StrategyDecision.CONSULT_REMAINING_STRATEGIES;
+            }
```
Comment (Vlatombe) [inline src/main/java/io/jenkins/plugins/kubernetes/NoDelayProvisionerStrategy.java:84]: So the key change resides here. If the Kubernetes strategy did not provision any node through `KubernetesCloud`, it goes to the next strategy, allowing other clouds to pick up the request.

However as soon as it was able to provision at least 1 node using a `KubernetesCloud`, we consider this label isn't served by other clouds so we just end the provisioning to prevent the standard strategy from causing overprovisioning.

**Comment by schottsfired** [comment]
Cool @Vlatombe. Interestingly I tried reverting https://github.com/jenkinsci/kubernetes-plugin/pull/824/commits/9b39286766d4818f4cad1d74946ec45c8c728a79 and committed again after changing perms to `744`, but `git` showed that perms were changed to `755`:
> ➜  kubernetes-plugin git:(fix-concurrency-limits) ✗ git commit -m "chmod 744 test-in-k8s.sh"
[fix-concurrency-limits 7233337f] chmod 744 test-in-k8s.sh
 1 file changed, 0 insertions(+), 0 deletions(-)
 mode change 100644 => 100755 test-in-k8s.sh

As per [SoF](https://stackoverflow.com/questions/21691202/how-to-create-file-execute-mode-permissions-in-git-on-windows) it seems `755` is the default, so the original commit seems OK as is.

**Review by Vlatombe — APPROVED** [review]

File: src/main/java/io/jenkins/plugins/kubernetes/NoDelayProvisionerStrategy.java (hunk 7 above)
Comment (schottsfired) [reply src/main/java/io/jenkins/plugins/kubernetes/NoDelayProvisionerStrategy.java:84]: Thanks for noting! I should have documented this in the PR description but I am not very familiar with Jenkins provisioning logic.

If a label is served by multiple strategies, and the Kubernetes strategy serves the first N requests up to the concurrency limit, will the remaining requests _eventually_ be served by other strategies?

Comment (Vlatombe) [reply src/main/java/io/jenkins/plugins/kubernetes/NoDelayProvisionerStrategy.java:84]: Not in the current round. If we reach the limit, on the next round, the Kubernetes strategy will provision 0 node and pass the ball to other strategies, however there won't be any more in-flight node because they get processed before each round of strategies (and the Kubernetes one returns instantly)

**Comment by wu105** [comment]
Is this supposed to reopen or referred to by a new issue when the fix is released with 1.27.1 but removed since 1.27.2?

**Comment by schottsfired** [comment]
Hi @wu105, this PR contains evidence of its reversion (https://github.com/jenkinsci/kubernetes-plugin/pull/844 above your comment), so I don't think it needs to be changed or reopened.

I do think that https://github.com/jenkinsci/kubernetes-plugin/pull/835 should be more strongly associated with [JENKINS-59959](https://issues.jenkins-ci.org/browse/JENKINS-59959), so that people who search for that issue ID find all relevant PRs, and not only this one.

cc @Vlatombe @jglick

**Comment by jglick** [comment]
@schottsfired feel free to add **Link**s in Jira.
